pkg/model/deployment_source.proto

```diff
@@ -0,0 +1,40 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package model;
+
+option go_package = "github.com/pipe-cd/pipecd/pkg/model";
+
+message DeploymentSource {
+  // The application directory where the source code is located.
+  string application_directory = 1;
+  // The git commit revision of the source code.
+  string revision = 2;
+  // The configuration of the application which is specific for plugins.
+  PluginApplicationSpec application_config = 3;
```

Comment:
I don't get the point of having this type of definition instead of just using bytes. Could you help me understand more about this 👀

Reply:
cc @Warashi

Reply:
I think kind and api_version exist for all configurations, so I defined them separately.
Additionally, these definitions reflect the limitation of the config.

Reply:
@Warashi Thanks for the explanation 🙏 I think this place we should not define a new type for lower maintenance costs. It could be better to just use bytes type here and both piped and plugin source code will use the `config.ApplicationConfig` type and marshall function defined in config package. wdyt? 🤔

Reply:
@khanhtc1202 
Do you mean [`config.Config`](https://github.com/pipe-cd/pipecd/blob/0e7a17704ab6c21895482826466cc02fc21ae457/pkg/config/config.go#L72-L89)?
I think the plugin source code will not use this because the config struct will be defined in the plugin code, not in the piped code.

Reply:
> I think this place we should not define a new type for lower maintenance costs.

I got your point 👍🏻 
I'll try to define a function or struct/method used by both piped and plugins.
I think this definition will use the generic type parameter.

Reply:
@khanhtc1202 
I fixed the type as `bytes` as you pointed.
In this PR, I want to define DeploymentSource only.
I'll implement config.ApplicationConfig or a similar one in another PR when I need it.

Reply:
@Warashi Thank you 👍